File: packages/android/captcha-compose/src/main/java/com/captcha/pro/compose/ClickCaptcha.kt
package com.captcha.pro.compose

import android.graphics.Bitmap
import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.scaleIn
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.graphics.drawscope.drawIntoCanvas
import androidx.compose.ui.graphics.nativeCanvas
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.unit.dp
import com.captcha.pro.core.BackendVerifyOptions
import com.captcha.pro.core.CaptchaGenerator
import com.captcha.pro.core.CaptchaLocale
import com.captcha.pro.core.CaptchaOptions
import com.captcha.pro.core.CaptchaPoint
import com.captcha.pro.core.CaptchaType
import com.captcha.pro.core.LocaleMessages
import com.captcha.pro.core.VerifyResult
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Click point data
 */
data class ClickPoint(
    val x: Float,
    val y: Float,
    val text: String?,
    val index: Int
)

/**
 * Click captcha composable — backend verification only.
 */
@Composable
fun ClickCaptcha(
    modifier: Modifier = Modifier,
    width: Int = 300,
    height: Int = 170,
    count: Int = 3,
    showRefresh: Boolean = true,
    backendVerify: BackendVerifyOptions,
    locale: CaptchaLocale = CaptchaLocale.ZH_CN,
    onSuccess: (VerifyResult?) -> Unit = {},
    onFail: () -> Unit = {},
    onRefresh: () -> Unit = {},
    onError: (Throwable) -> Unit = {},
) {
    val generator = remember { CaptchaGenerator() }
    val scope = rememberCoroutineScope()

    var bgBitmap by remember { mutableStateOf<Bitmap?>(null) }
    var clickTexts by remember { mutableStateOf<List<String>>(emptyList()) }
    var clickCharImages by remember { mutableStateOf<List<String>>(emptyList()) }
    var charBitmaps by remember { mutableStateOf<List<Bitmap>>(emptyList()) }
    var clickPoints by remember { mutableStateOf<List<ClickPoint>>(emptyList()) }
    var status by remember { mutableStateOf<String?>(null) }
    var loading by remember { mutableStateOf(true) }
    var errorMsg by remember { mutableStateOf<String?>(null) }

    val gradientColors = listOf(Color(0xFF667EEA), Color(0xFF764BA2))

    val targetCount = when {
        clickCharImages.isNotEmpty() -> clickCharImages.size
        clickTexts.isNotEmpty() -> clickTexts.size
        else -> count
    }

    fun refresh() {
        scope.launch {
            loading = true
            errorMsg = null
            clickPoints = emptyList()
            try {
                val result = generator.generate(
                    CaptchaOptions(
                        type = CaptchaType.CLICK,
                        width = width,
                        height = height,
                        clickCount = count,
                        backendVerify = backendVerify,
                        locale = locale
                    )
                )
                bgBitmap = result.bgBitmap
                clickTexts = result.clickTexts ?: emptyList()
                clickCharImages = result.clickCharImages ?: emptyList()
                status = null
                loading = false
                onRefresh()

                // Load char images off the main thread when present.
                charBitmaps = if (clickCharImages.isNotEmpty()) {
                    withContext(Dispatchers.IO) {
                        clickCharImages.mapNotNull { src -> runCatching { generator.loadImage(src) }.getOrNull() }
                    }
                } else emptyList()
            } catch (e: Exception) {
                loading = false
                errorMsg = LocaleMessages.get(locale, "error_network")
                onError(e)
            }
        }
    }

    fun handleClick(offsetX: Float, offsetY: Float) {
        if (status != null || loading) return
        if (clickPoints.size >= targetCount) return

        val newClickPoints = clickPoints + ClickPoint(offsetX, offsetY, null, clickPoints.size)
        clickPoints = newClickPoints

        if (newClickPoints.size == targetCount) {
            val points = newClickPoints.map { CaptchaPoint(it.x, it.y) }
            val data = generator.getCaptchaData(type = CaptchaType.CLICK, targetPoints = points)
            val options = CaptchaOptions(type = CaptchaType.CLICK, backendVerify = backendVerify, locale = locale)
            scope.launch {
                try {
                    val response = generator.backendVerify(data, options)
                    if (response.success) {
                        status = "success"
                        val verifiedAt = (response.data?.get("verifiedAt") as? Number)?.toLong()
                        onSuccess(VerifyResult(verifiedAt = verifiedAt))
                    } else {
                        status = "fail"
                        onFail()
                        delay(800)
                        refresh()
                    }
                } catch (e: Exception) {
                    status = "fail"
                    errorMsg = LocaleMessages.get(locale, "error_network")
                    onError(e)
                }
            }
        }
    }

    LaunchedEffect(Unit) { refresh() }

    Column(modifier = modifier.width(width.dp)) {
        // Captcha area
        Box(
            modifier = Modifier
                .width(width.dp)
                .height(height.dp)
                .shadow(8.dp, RoundedCornerShape(16.dp))
                .clip(RoundedCornerShape(16.dp))
                .background(Brush.linearGradient(gradientColors))
        ) {
            // Background image
            bgBitmap?.let { bitmap ->
                Canvas(
                    modifier = Modifier
                        .matchParentSize()
                        .pointerInput(Unit) {
                            detectTapGestures { offset ->
                                handleClick(offset.x, offset.y)
                            }
                        }
                ) {
                    drawIntoCanvas { canvas ->
                        canvas.nativeCanvas.drawBitmap(
                            bitmap,
                            null,
                            android.graphics.Rect(0, 0, size.width.toInt(), size.height.toInt()),
                            null
                        )
                    }
                }
            } ?: Text(
                text = errorMsg ?: LocaleMessages.get(locale, "loading"),
                color = Color.White,
                modifier = Modifier.align(Alignment.Center)
            )

            // Click point indicators
            clickPoints.forEach { point ->
                Box(
                    modifier = Modifier
                        .offset(x = (point.x - 12).dp, y = (point.y - 12).dp)
                        .size(24.dp)
                        .background(Color(0xFF1991FA), CircleShape)
                        .border(3.dp, Color.White, CircleShape),
                    contentAlignment = Alignment.Center
                ) {
                    Text(
                        text = "${point.index + 1}",
                        color = Color.White,
                        style = MaterialTheme.typography.caption
                    )
                }
            }

            // Refresh button
            if (showRefresh && !loading) {
                Box(
                    modifier = Modifier
                        .align(Alignment.TopEnd)
                        .padding(8.dp)
                        .size(28.dp)
                        .clip(RoundedCornerShape(8.dp))
                        .background(Color.White.copy(alpha = 0.9f))
                        .clickable { refresh() },
                    contentAlignment = Alignment.Center
                ) {
                    Text("⟳", color = Color(0xFF666666))
                }
            }

            // Status overlay — centered, white@75%, animated fade-in + scale
            androidx.compose.animation.AnimatedVisibility(
                visible = status != null,
                enter = fadeIn(animationSpec = tween(200)) +
                    scaleIn(initialScale = 0.9f, animationSpec = tween(200)),
                exit = fadeOut(animationSpec = tween(200))
            ) {
                val s = status ?: ""
                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .background(Color.White.copy(alpha = 0.75f)),
                    contentAlignment = Alignment.Center
                ) {
                    Column(horizontalAlignment = Alignment.CenterHorizontally) {
                        Box(
                            modifier = Modifier
                                .size(32.dp)
                                .clip(CircleShape)
                                .background(
                                    if (s == "success") Color(0xFF52C41A).copy(alpha = 0.85f)
                                    else Color(0xFFFF4D4F).copy(alpha = 0.85f)
                                ),
                            contentAlignment = Alignment.Center
                        ) {
                            Text(
                                text = if (s == "success") "✓" else "✕",
                                color = Color.White,
                                style = MaterialTheme.typography.h6
                            )
                        }
                        Spacer(modifier = Modifier.height(8.dp))
                        Text(
                            text = if (s == "success")
                                LocaleMessages.get(locale, "click_success")
                            else LocaleMessages.get(locale, "click_fail"),
                            color = if (s == "success") Color(0xFF389E0D) else Color(0xFFCF1322)
                        )
                    }
                }
            }
        }

        Spacer(modifier = Modifier.height(12.dp))

        // Prompt bar
        Surface(
            modifier = Modifier.fillMaxWidth(),
            color = Color(0xFFF7F9FA),
            shape = RoundedCornerShape(16.dp),
            border = BorderStroke(1.dp, Color(0xFFE8E8E8))
        ) {
            Row(
                modifier = Modifier.padding(horizontal = 12.dp, vertical = 10.dp),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Row(
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text(
                        text = LocaleMessages.get(locale, "click_prompt"),
                        color = Color(0xFF666666),
                        style = MaterialTheme.typography.body2
                    )
                    Spacer(modifier = Modifier.width(8.dp))
                    if (charBitmaps.isNotEmpty()) {
                        charBitmaps.forEachIndexed { i, bitmap ->
                            Box(
                                modifier = Modifier
                                    .padding(start = 6.dp)
                                    .size(28.dp)
                                    .shadow(2.dp, RoundedCornerShape(8.dp))
                                    .clip(RoundedCornerShape(8.dp))
                                    .background(Brush.linearGradient(gradientColors)),
                                contentAlignment = Alignment.Center
                            ) {
                                Image(
                                    bitmap = bitmap.asImageBitmap(),
                                    contentDescription = "char${i + 1}",
                                    modifier = Modifier.fillMaxSize().padding(4.dp)
                                )
                            }
                        }
                    } else {
                        clickTexts.forEach { char ->
                            Box(
                                modifier = Modifier
                                    .padding(start = 6.dp)
                                    .size(28.dp)
                                    .shadow(2.dp, RoundedCornerShape(8.dp))
                                    .clip(RoundedCornerShape(8.dp))
                                    .background(Brush.linearGradient(gradientColors)),
                                contentAlignment = Alignment.Center
                            ) {
                                Text(
                                    char,
                                    color = Color.White,
                                    style = MaterialTheme.typography.body1
                                )
                            }
                        }
                    }
                }
                Text(
                    text = "${clickPoints.size}/$targetCount",
                    style = MaterialTheme.typography.body2,
                    color = Color(0xFF1991FA)
                )
            }
        }
    }
}

File: packages/ios/Sources/SwiftUI/ClickCaptcha.swift
import SwiftUI
import UIKit

/// Click captcha view for SwiftUI — backend verification only.
@available(iOS 14.0, *)
public struct ClickCaptcha: View {
    @StateObject private var viewModel: ClickCaptchaViewModel

    public init(
        width: Int = 300,
        height: Int = 170,
        count: Int = 3,
        showRefresh: Bool = true,
        backendVerify: BackendVerifyOptions,
        locale: CaptchaLocale = .zhCN,
        onSuccess: @escaping (VerifyResult?) -> Void = { _ in },
        onFail: @escaping () -> Void = {},
        onRefresh: @escaping () -> Void = {},
        onError: @escaping (Error) -> Void = { _ in }
    ) {
        _viewModel = StateObject(wrappedValue: ClickCaptchaViewModel(
            width: width,
            height: height,
            count: count,
            showRefresh: showRefresh,
            backendVerify: backendVerify,
            locale: locale,
            onSuccess: onSuccess,
            onFail: onFail,
            onRefresh: onRefresh,
            onError: onError
        ))
    }

    public var body: some View {
        VStack(spacing: 12) {
            // Captcha area
            ZStack(alignment: .topTrailing) {
                LinearGradient(
                    colors: [
                        Color(red: 102/255, green: 126/255, blue: 234/255),
                        Color(red: 118/255, green: 75/255, blue: 162/255),
                    ],
                    startPoint: .topLeading,
                    endPoint: .bottomTrailing
                )

                if let bgImage = viewModel.bgImage {
                    Image(uiImage: bgImage)
                        .resizable()
                        .frame(width: CGFloat(viewModel.width), height: CGFloat(viewModel.height))
                        .contentShape(Rectangle())
                        .gesture(
                            DragGesture(minimumDistance: 0)
                                .onEnded { value in
                                    viewModel.handleClick(at: value.location)
                                }
                        )
                        .accessibilityLabel(LocaleMessages.get(viewModel.locale, key: "click_prompt"))
                        .accessibilityAddTraits(.isButton)
                } else if let msg = viewModel.errorMsg {
                    Text(msg)
                        .foregroundColor(.white)
                        .font(.system(size: 14))
                } else {
                    Text(LocaleMessages.get(viewModel.locale, key: "loading"))
                        .foregroundColor(.white)
                        .font(.system(size: 14))
                }

                // Click markers
                ForEach(viewModel.clickPoints.indices, id: \.self) { index in
                    let point = viewModel.clickPoints[index]
                    Circle()
                        .fill(Color(red: 25/255, green: 145/255, blue: 250/255))
                        .frame(width: 24, height: 24)
                        .overlay(Circle().stroke(Color.white, lineWidth: 3))
                        .overlay(
                            Text("\(index + 1)")
                                .foregroundColor(.white)
                                .font(.system(size: 12))
                                .fontWeight(.bold)
                        )
                        .offset(x: point.x - 12, y: point.y - 12)
                        .accessibilityLabel("\(index + 1)")
                }

                // Refresh button
                if viewModel.showRefresh, !viewModel.loading {
                    Button(action: { viewModel.refresh() }) {
                        Text("⟳")
                            .foregroundColor(Color(white: 0.4))
                            .font(.system(size: 20, weight: .bold))
                            .frame(width: 28, height: 28)
                            .background(Color.white.opacity(0.9))
                            .cornerRadius(8)
                    }
                    .padding(8)
                }

                // Status overlay (centered)
                if let status = viewModel.status {
                    VStack(spacing: 12) {
                        ZStack {
                            Circle()
                                .fill(status == .success
                                      ? Color(red: 82/255, green: 196/255, blue: 26/255, opacity: 0.85)
                                      : Color(red: 255/255, green: 77/255, blue: 79/255, opacity: 0.85))
                                .frame(width: 32, height: 32)
                            Text(status == .success ? "✓" : "✕")
                                .foregroundColor(.white)
                                .font(.system(size: 18, weight: .bold))
                        }
                        Text(status == .success
                             ? LocaleMessages.get(viewModel.locale, key: "click_success")
                             : LocaleMessages.get(viewModel.locale, key: "click_fail"))
                            .font(.system(size: 14, weight: .medium))
                            .foregroundColor(status == .success
                                             ? Color(red: 56/255, green: 158/255, blue: 13/255)
                                             : Color(red: 207/255, green: 19/255, blue: 34/255))
                    }
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
                    .background(Color.white.opacity(0.75))
                    .transition(.opacity.combined(with: .scale(scale: 0.9)))
                }
            }
            .frame(width: CGFloat(viewModel.width), height: CGFloat(viewModel.height))
            .cornerRadius(16)
            .shadow(color: Color.black.opacity(0.15), radius: 16, x: 0, y: 4)
            .animation(.easeInOut(duration: 0.2), value: viewModel.status)

            // Prompt bar
            HStack(spacing: 8) {
                Text(LocaleMessages.get(viewModel.locale, key: "click_prompt"))
                    .font(.system(size: 14))
                    .foregroundColor(Color(white: 0.4))

                if !viewModel.clickCharImages.isEmpty {
                    HStack(spacing: 6) {
                        ForEach(viewModel.clickCharImages.indices, id: \.self) { i in
                            if i < viewModel.charImages.count {
                                let img = viewModel.charImages[i]
                                Image(uiImage: img)
                                    .resizable()
                                    .frame(width: 28, height: 28)
                                    .clipShape(RoundedRectangle(cornerRadius: 8))
                                    .overlay(
                                        RoundedRectangle(cornerRadius: 8)
                                            .stroke(
                                                LinearGradient(
                                                    colors: [
                                                        Color(red: 102/255, green: 126/255, blue: 234/255),
                                                        Color(red: 118/255, green: 75/255, blue: 162/255),
                                                    ],
                                                    startPoint: .topLeading,
                                                    endPoint: .bottomTrailing
                                                ),
                                                lineWidth: 1
                                            )
                                    )
                            } else {
                                LinearGradient(
                                    colors: [
                                        Color(red: 102/255, green: 126/255, blue: 234/255),
                                        Color(red: 118/255, green: 75/255, blue: 162/255),
                                    ],
                                    startPoint: .topLeading,
                                    endPoint: .bottomTrailing
                                )
                                .frame(width: 28, height: 28)
                                .cornerRadius(8)
                            }
                        }
                    }
                } else {
                    HStack(spacing: 6) {
                        ForEach(viewModel.clickTexts.indices, id: \.self) { i in
                            LinearGradient(
                                colors: [
                                    Color(red: 102/255, green: 126/255, blue: 234/255),
                                    Color(red: 118/255, green: 75/255, blue: 162/255),
                                ],
                                startPoint: .topLeading,
                                endPoint: .bottomTrailing
                            )
                            .frame(width: 28, height: 28)
                            .cornerRadius(8)
                            .shadow(color: Color(red: 102/255, green: 126/255, blue: 234/255, opacity: 0.3), radius: 4)
                            .overlay(
                                Text(viewModel.clickTexts[i])
                                    .foregroundColor(.white)
                                    .font(.system(size: 16, weight: .medium))
                            )
                        }
                    }
                }

                Spacer()

                Text("\(viewModel.clickPoints.count)/\(viewModel.targetCount)")
                    .font(.system(size: 14))
                    .foregroundColor(Color(red: 25/255, green: 145/255, blue: 250/255))
            }
            .padding(.horizontal, 12)
            .frame(width: CGFloat(viewModel.width), height: 56)
            .background(Color(red: 247/255, green: 249/255, blue: 250/255))
            .cornerRadius(16)
            .overlay(
                RoundedRectangle(cornerRadius: 16)
                    .stroke(Color(red: 232/255, green: 232/255, blue: 232/255), lineWidth: 1)
            )
        }
        .onAppear { viewModel.refresh() }
    }
}

/// Click captcha view model — backend verification only.
@MainActor
final class ClickCaptchaViewModel: ObservableObject {
    @Published var bgImage: UIImage?
    @Published var clickPoints: [CGPoint] = []
    @Published var clickTexts: [String] = []
    @Published var clickCharImages: [String] = []
    @Published var charImages: [UIImage] = []
    @Published var status: Status?
    @Published var loading = false
    @Published var errorMsg: String?

    private let generator = CaptchaGenerator()

    let width: Int
    let height: Int
    let count: Int
    let showRefresh: Bool
    let backendVerify: BackendVerifyOptions
    let locale: CaptchaLocale
    let onSuccess: (VerifyResult?) -> Void
    let onFail: () -> Void
    let onRefresh: () -> Void
    let onError: (Error) -> Void

    enum Status: Equatable {
        case success, fail
    }

    var targetCount: Int {
        if !clickCharImages.isEmpty { return clickCharImages.count }
        if !clickTexts.isEmpty { return clickTexts.count }
        return count
    }

    init(
        width: Int,
        height: Int,
        count: Int,
        showRefresh: Bool,
        backendVerify: BackendVerifyOptions,
        locale: CaptchaLocale,
        onSuccess: @escaping (VerifyResult?) -> Void,
        onFail: @escaping () -> Void,
        onRefresh: @escaping () -> Void,
        onError: @escaping (Error) -> Void
    ) {
        self.width = width
        self.height = height
        self.count = count
        self.showRefresh = showRefresh
        self.backendVerify = backendVerify
        self.locale = locale
        self.onSuccess = onSuccess
        self.onFail = onFail
        self.onRefresh = onRefresh
        self.onError = onError
    }

    func refresh() {
        clickPoints.removeAll()
        loading = true
        errorMsg = nil
        let options = CaptchaOptions(
            type: .click,
            width: width,
            height: height,
            clickCount: count,
            backendVerify: backendVerify,
            locale: locale
        )
        Task {
            do {
                let result = try await generator.generate(options: options)
                bgImage = result.bgImage
                clickTexts = result.clickTexts ?? []
                clickCharImages = result.clickCharImages ?? []
                status = nil
                loading = false
                onRefresh()

                // Load char images off the main actor when present.
                if !clickCharImages.isEmpty {
                    var images: [UIImage?] = Array(repeating: nil, count: clickCharImages.count)
                    for (i, src) in clickCharImages.enumerated() {
                        if let img = try? await generator.loadImage(from: src) {
                            images[i] = img
                        }
                    }
                    charImages = images.compactMap { $0 }
                } else {
                    charImages = []
                }
            } catch {
                loading = false
                errorMsg = LocaleMessages.get(locale, key: "error_network")
                onError(error)
            }
        }
    }

    func handleClick(at location: CGPoint) {
        guard status == nil, clickPoints.count < targetCount else { return }
        clickPoints.append(location)
        if clickPoints.count == targetCount {
            verify()
        }
    }

    private func verify() {
        let points = clickPoints.map { CaptchaPoint(x: $0.x, y: $0.y) }
        let data = generator.getCaptchaData(type: .click, targetPoints: points)
        let options = CaptchaOptions(type: .click, backendVerify: backendVerify, locale: locale)
        Task {
            do {
                let response = try await generator.backendVerify(data: data, options: options)
                if response.success {
                    status = .success
                    onSuccess(VerifyResult(verifiedAt: response.verifiedAt))
                } else {
                    status = .fail
                    onFail()
                    DispatchQueue.main.asyncAfter(deadline: .now() + 0.8) { [weak self] in
                        self?.refresh()
                    }
                }
            } catch {
                status = .fail
                errorMsg = LocaleMessages.get(locale, key: "error_network")
                onError(error)
            }
        }
    }
}
